[Profile] Add binary profile correlation for code coverage.

ZequanWu opened the pull request:
## Motivation
Since we don't need the metadata sections at runtime, we can somehow offload them from memory at runtime. Initially, I explored [debug info correlation](https://discourse.llvm.org/t/instrprofiling-lightweight-instrumentation/59113), which is used for PGO with value profiling disabled. However, it currently only works with DWARF and it's be hard to add such artificial debug info for every function in to CodeView which is used on Windows. So, offloading profile metadata sections at runtime seems to be a platform independent option.

## Design
The idea is to use new section names for profile name and data sections and mark them as metadata sections. Under this mode, the new sections are non-SHF_ALLOC in ELF. So, they are not loaded into memory at runtime and can be stripped away as a post-linking step. After the process exits, the generated raw profiles will contains only headers + counters. llvm-profdata can be used correlate raw profiles with the unstripped binary to generate indexed profile.

## Data
For chromium base_unittests with code coverage on linux, the binary size overhead due to instrumentation reduced from 64M to 38.8M (39.4%) and the raw profile files size reduce from 128M to 68M (46.9%)
```
$ bloaty out/cov/base_unittests.stripped -- out/no-cov/base_unittests.stripped
    FILE SIZE        VM SIZE
 --------------  --------------
  +121% +30.4Mi  +121% +30.4Mi    .text
  [NEW] +14.6Mi  [NEW] +14.6Mi    __llvm_prf_data
  [NEW] +10.6Mi  [NEW] +10.6Mi    __llvm_prf_names
  [NEW] +5.86Mi  [NEW] +5.86Mi    __llvm_prf_cnts
   +95% +1.75Mi   +95% +1.75Mi    .eh_frame
  +108%  +400Ki  +108%  +400Ki    .eh_frame_hdr
  +9.5%  +211Ki  +9.5%  +211Ki    .rela.dyn
  +9.2% +95.0Ki  +9.2% +95.0Ki    .data.rel.ro
  +5.0% +87.3Ki  +5.0% +87.3Ki    .rodata
  [ = ]       0   +13% +47.0Ki    .bss
   +40% +1.78Ki   +40% +1.78Ki    .got
   +12% +1.49Ki   +12% +1.49Ki    .gcc_except_table
  [ = ]       0   +65% +1.23Ki    .relro_padding
   +62% +1.20Ki  [ = ]       0    [Unmapped]
   +13%    +448   +19%    +448    .init_array
  +8.8%    +192  [ = ]       0    [ELF Section Headers]
  +0.0%    +136  +0.0%     +80    [7 Others]
  +0.1%     +96  +0.1%     +96    .dynsym
  +1.2%     +96  +1.2%     +96    .rela.plt
  +1.5%     +80  +1.2%     +64    .plt
  [ = ]       0 -99.2% -3.68Ki    [LOAD #5 [RW]]
  +195% +64.0Mi  +194% +64.0Mi    TOTAL
$ bloaty out/cov-cor/base_unittests.stripped -- out/no-cov/base_unittests.stripped
    FILE SIZE        VM SIZE
 --------------  --------------
  +121% +30.4Mi  +121% +30.4Mi    .text
  [NEW] +5.86Mi  [NEW] +5.86Mi    __llvm_prf_cnts
   +95% +1.75Mi   +95% +1.75Mi    .eh_frame
  +108%  +400Ki  +108%  +400Ki    .eh_frame_hdr
  +9.5%  +211Ki  +9.5%  +211Ki    .rela.dyn
  +9.2% +95.0Ki  +9.2% +95.0Ki    .data.rel.ro
  +5.0% +87.3Ki  +5.0% +87.3Ki    .rodata
  [ = ]       0   +13% +47.0Ki    .bss
   +40% +1.78Ki   +40% +1.78Ki    .got
   +12% +1.49Ki   +12% +1.49Ki    .gcc_except_table
   +13%    +448   +19%    +448    .init_array
  +0.1%     +96  +0.1%     +96    .dynsym
  +1.2%     +96  +1.2%     +96    .rela.plt
  +1.2%     +64  +1.2%     +64    .plt
  +2.9%     +64  [ = ]       0    [ELF Section Headers]
  +0.0%     +40  +0.0%     +40    .data
  +1.2%     +32  +1.2%     +32    .got.plt
  +0.0%     +24  +0.0%      +8    [5 Others]
  [ = ]       0 -22.9%    -872    [LOAD #5 [RW]]
 -74.5% -1.44Ki  [ = ]       0    [Unmapped]
  [ = ]       0 -76.5% -1.45Ki    .relro_padding
  +118% +38.8Mi  +117% +38.8Mi    TOTAL
```

A few things to note:
1. llvm-profdata doesn't support filter raw profiles by binary id yet, so when a raw profile doesn't belongs to the binary being digested by llvm-profdata, merging will fail. Once this is implemented, llvm-profdata should be able to only merge raw profiles with the same binary id as the binary and discard the rest (with mismatched/missing binary id). The workflow I have in mind is to have scripts invoke llvm-profdata to get all binary ids for all raw profiles, and selectively choose the raw pnrofiles with matching binary id and the binary to llvm-profdata for merging.
2. Note: In COFF, currently they are still loaded into memory but not used. I didn't do it in this patch because I noticed that `.lcovmap` and `.lcovfunc` are loaded into memory. A separate patch will address it.
3. This should works with PGO when value profiling is disabled as debug info correlation currently doing, though I haven't tested this yet.

ZequanWu committed:
[Profile] Add binary profile correlation.

ZequanWu committed:
fix correlator for coff.

ZequanWu committed:
update docs and renaming option

ZequanWu committed:
format

ZequanWu committed:
add llvm-cov and IR test

ZequanWu commented on llvm/test/Instrumentation/InstrProfiling/binary-correlate.ll:
```diff
@@ -0,0 +1,11 @@
+; RUN: opt < %s -passes=instrprof -profile-correlate=binary -S | FileCheck %s
+
+; CHECK: @__profd_foo = private global { i64, i64, i64, ptr, ptr, i32, [2 x i16] } { i64 {{.*}}, i64 {{.*}}, i64 ptrtoint (ptr @__profc_foo to i64)
```
Though the CounterPtr's type (3rd argument) is i64, not ptr, I found that when lowered to assembly, it is just the address of the counter label. I guess it's fine.

ellishg commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -60,10 +60,7 @@ using namespace llvm;
 #define DEBUG_TYPE "instrprof"
 
 namespace llvm {
-cl::opt<bool>
-    DebugInfoCorrelate("debug-info-correlate",
-                       cl::desc("Use debug info to correlate profiles."),
-                       cl::init(false));
```
Is it possible to make this an alias to `-profile-correlate=debug-info`? I do like the new flag better, but I want to make sure we don't break existing code that relies on this flag. We should give users a chance to migrate.

ZequanWu replied:
I found cl::alias cannot be used to alias one option of another flag (e.g. -profile-correlate=debug-info). So, every place we uses ProfileCorrelate needs to have a check for DebugInfoCorrelate. I don't know how you update llvm toolchain, is it possible to switch to new flag at the time when updating the toolchain?

ZequanWu replied:
I refactored some code to #69656 and kept this flag `-debug-info-correlate`.

david-xl commented:
Can you send a RFC for this enhancement ? The first 2 paragraph of this PR can be expanded a little more to show motivations (RAM saving, or raw profile size saving and why) and the savings data.

zmodem reviewed (commented):
I think there should be a proper Clang flag to control this, instead of requiring users to pass internal `-mllvm` flags. (Also the flag should be well documented.)

rnk commented:
> I think there should be a proper Clang flag to control this, instead of requiring users to pass internal `-mllvm` flags. (Also the flag should be well documented.)

I believe Zequan did that at one point, but he ran into the problem that Rust and Clang need to agree on the coverage metadata section flags. Using an mllvm flag means we can opt into the new mode for both clang and rust simultaneously, hopefully without waiting for rustc changes.

It's interesting that historically we have discouraged mllvm flags, but in a new world of multiple frontends outside the llvm-project monorepo, having a direct command line interface to the middle-end becomes more valuable.

david-xl commented on clang/lib/CodeGen/BackendUtil.cpp:
```diff
@@ -98,13 +99,16 @@ extern cl::opt<bool> PrintPipelinePasses;
     cl::desc("Insert sanitizers on OptimizerEarlyEP."), cl::init(false));
-}
+
+extern cl::opt<InstrProfCorrelator::ProfCorrelatorKind> ProfileCorrelate;
```
Can be extracted out as refactoring (without adding the new correlation kind).

david-xl replied:
yes - just use the debug-info kind and non-kind.

ZequanWu replied:
There are three options to ProfileCorrelate, none (default), debug-info and binary. This patch introduces the binary mode, which is controlled by -mllvm -profile-correlate=binary. And both binary and debug-info correlation will have the default profile name `default_%m.proflite`.

david-xl replied:
right -- what I meant is that you can move this option into the refactor patch. In that patch, the option will only have two options.

ZequanWu replied:
Oh, I see. Do you mean remove ProfCorrelatorKind::None? Just one for debug-info and another one for binary?

david-xl replied:
No, what I meant is in the refactoring/preparation patch, change debug-info-correlate into profile-correlate=debug. The none kind is still kept. 

In your next patch to implement binary correlation, add the binary kind.

The goal for the preparation/refactoring patch is to minimize the diff in the second patch implementing the feature so that reviewer can have more focus on the functional diffs.

david-xl commented on compiler-rt/lib/profile/InstrProfiling.h:
```diff
@@ -267,6 +267,9 @@ uint64_t __llvm_profile_get_num_data(const __llvm_profile_data *Begin,
                                       const __llvm_profile_data *End);
 
+/*! \brief Get the size of the profile name section in bytes. */
+uint64_t __llvm_profile_get_name_size(const char *Begin, const char *End);
```
extract out as in a refactoring patch. If this interface is not intended to be public, no need to use the __llvm name space for it, but use the internal interface naming convention.

david-xl commented on compiler-rt/lib/profile/InstrProfilingBuffer.c:
```diff
@@ -53,6 +53,9 @@ uint64_t __llvm_profile_get_size_for_buffer(void) {
 COMPILER_RT_VISIBILITY
 uint64_t __llvm_profile_get_num_data(const __llvm_profile_data *Begin,
                                      const __llvm_profile_data *End) {
+  if ((__llvm_profile_get_version() & VARIANT_MASK_DBG_CORRELATE) ||
```
Add another inline helper : hasCorrelation(..) and put it in the refactor patch.

david-xl commented on compiler-rt/lib/profile/InstrProfilingBuffer.c:
```diff
@@ -61,9 +64,20 @@ uint64_t __llvm_profile_get_num_data(const __llvm_profile_data *Begin,
 COMPILER_RT_VISIBILITY
 uint64_t __llvm_profile_get_data_size(const __llvm_profile_data *Begin,
                                       const __llvm_profile_data *End) {
+  if ((__llvm_profile_get_version() & VARIANT_MASK_DBG_CORRELATE) ||
```
same here.

david-xl commented on compiler-rt/lib/profile/InstrProfilingBuffer.c:
```diff
@@ -61,9 +64,20 @@ uint64_t __llvm_profile_get_num_data(const __llvm_profile_data *Begin,
 
+COMPILER_RT_VISIBILITY
+uint64_t __llvm_profile_get_name_size(const char *Begin, const char *End) {
+  if ((__llvm_profile_get_version() & VARIANT_MASK_DBG_CORRELATE) ||
```
here

david-xl commented on compiler-rt/lib/profile/InstrProfilingMerge.c:
```diff
@@ -66,8 +66,9 @@ int __llvm_profile_check_compatibility(const char *ProfileData,
                                           __llvm_profile_end_counters()) ||
-      Header->NamesSize != (uint64_t)(__llvm_profile_end_names() -
-                                      __llvm_profile_begin_names()) ||
+      Header->NamesSize !=
```
candidate in refactor patch

david-xl commented on compiler-rt/lib/profile/InstrProfilingMerge.c:
```diff
@@ -132,7 +133,8 @@ int __llvm_profile_merge_from_buffer(const char *ProfileData,
   // Merge counters by iterating the entire counter section when debug info
   // correlation is enabled.
-  if (__llvm_profile_get_version() & VARIANT_MASK_DBG_CORRELATE) {
+  if ((__llvm_profile_get_version() & VARIANT_MASK_DBG_CORRELATE) ||
```
use helper function and extract it to refactor patch.

david-xl commented on compiler-rt/lib/profile/InstrProfilingWriter.c:
```diff
@@ -259,19 +259,19 @@ lprofWriteDataImpl(ProfDataWriter *Writer, const __llvm_profile_data *DataBegin,
-  int DebugInfoCorrelate =
-      (__llvm_profile_get_version() & VARIANT_MASK_DBG_CORRELATE) != 0ULL;
+  int ProfileCorrelation =
+      (__llvm_profile_get_version() & VARIANT_MASK_DBG_CORRELATE) ||
```
put in refactor patch.

david-xl commented on compiler-rt/lib/profile/InstrProfilingWriter.c:
```diff
@@ -259,19 +259,19 @@ lprofWriteDataImpl(ProfDataWriter *Writer, const __llvm_profile_data *DataBegin,
-  const uint64_t NumData =
-      DebugInfoCorrelate ? 0 : __llvm_profile_get_num_data(DataBegin, DataEnd);
+      __llvm_profile_get_data_size(DataBegin, DataEnd);
+  const uint64_t NumData = __llvm_profile_get_num_data(DataBegin, DataEnd);
```
same here

david-xl commented on llvm/include/llvm/ProfileData/InstrProfReader.h:
```diff
@@ -374,6 +376,14 @@ class RawInstrProfReader : public InstrProfReader {
+    return (Version & VARIANT_MASK_BIN_CORRELATE) != 0;
+  }
+
+  bool useCorrelate() const {
```
candidate in refactoring patch.

david-xl commented on llvm/lib/ProfileData/InstrProfReader.cpp:
```diff
@@ -540,10 +540,10 @@ Error RawInstrProfReader<IntPtrT>::readHeader(
-    return error(instrprof_error::missing_debug_info_for_correlation);
-  if (!useDebugInfoCorrelate() && Correlator)
-    return error(instrprof_error::unexpected_debug_info_for_correlation);
+  if (useCorrelate() && !Correlator)
```
candidate refactoring change.

ZequanWu commented:
> > I think there should be a proper Clang flag to control this, instead of requiring users to pass internal `-mllvm` flags. (Also the flag should be well documented.)
> 
> I believe Zequan did that at one point, but he ran into the problem that Rust and Clang need to agree on the coverage metadata section flags. Using an mllvm flag means we can opt into the new mode for both clang and rust simultaneously, hopefully without waiting for rustc changes.

Since I modified clang CoverageMappingGen.cpp to conditionally add toggle this new mode in __llvm_profile_raw_version variable, rustc needs to do similar change at frontend in order to use this with code coverage. 

The problem I ran into is chromium unittests have rust code. When I tried this mode for base_unittests, the Rust binaries are not able to use this mode while the C/C++ binaries uses it. So, the generated raw profiles are mixture of Rust raw profiles (counters + metadata sections) + C/C++ raw profiles (counters). llvm-profdata is currently not able to handle this. 
My plan is to fix this in llvm-profdata later to allow merging with mixture raw profiles.

I'm not sure if we can pass -mllvm flag to rustc.

zmodem commented:
> > > I think there should be a proper Clang flag to control this, instead of requiring users to pass internal `-mllvm` flags. (Also the flag should be well documented.)
> > 
> > 
> > I believe Zequan did that at one point, but he ran into the problem that Rust and Clang need to agree on the coverage metadata section flags. Using an mllvm flag means we can opt into the new mode for both clang and rust simultaneously, hopefully without waiting for rustc changes.
> 
> Since I modified clang CoverageMappingGen.cpp to conditionally add toggle this new mode in __llvm_profile_raw_version variable, rustc needs to do similar change at frontend in order to use this with code coverage.

Interesting. Yes, this does seem like a good motivation for using internal flags, but I think the long-term aim should still be that we get proper user-facing flags in both clang and rustc.

> I'm not sure if we can pass -mllvm flag to rustc.

I think the equivalent of `-mllvm x` is `-C llvm-args=x` for rustc.

WenleiHe commented:
> This should works with PGO when value profiling is disabled 

Is this not compatible with value profiling? or just not implemented yet?

david-xl commented:
> > This should works with PGO when value profiling is disabled
> 
> Is this not compatible with value profiling? or just not implemented yet?

Not compatible, but this feature is mainly for coverage testing.

WenleiHe commented:
> > > This should works with PGO when value profiling is disabled
> > 
> > 
> > Is this not compatible with value profiling? or just not implemented yet?
> 
> Not compatible, but this feature is mainly for coverage testing.

Ok, not a big deal as it's probably not worth the effort. But it would be nice if metadata can be separated out orthogonal to functionality, so things work in unified way with less special case. Just like debug info can be in a standalone binary (with gnu_debuglink), or in the original ELF binary, and they work equivalently for symbolization.

ZequanWu commented:
> > > > This should works with PGO when value profiling is disabled
> > > 
> > > 
> > > Is this not compatible with value profiling? or just not implemented yet?
> > 
> > 
> > Not compatible, but this feature is mainly for coverage testing.
> 
> Ok, not a big deal as it's probably not worth the effort. But it would be nice if metadata can be separated out orthogonal to functionality, so things work in unified way with less special case. Just like debug info can be in a standalone binary (with gnu_debuglink), or in the original ELF binary, and they work equivalently for symbolization.

Ideally, prof name section in PGO can also be separated out as a strippable section. If we plan to separate PGO and code coverage stuff under compiler-rt/profile in the future. This is a good candidate as part of that effort.

ZequanWu committed:
resolve conflict

ZequanWu committed:
resolve conflict

ZequanWu committed:
add missing DebugInfoCorrelate

MaskRay commented on clang/lib/CodeGen/CoverageMappingGen.cpp:
```diff
@@ -1829,6 +1833,22 @@ void CoverageMappingModuleGen::emit() {
+      VarName);
+  VersionVariable->setVisibility(llvm::GlobalValue::HiddenVisibility);
+  llvm::Triple TT(CGM.getModule().getTargetTriple());
+  if (TT.supportsCOMDAT()) {
```
`CGM.getTriple().supportsCOMDAT()`

MaskRay commented on clang/test/CodeGen/coverage-profile-raw-version.c:
```diff
@@ -0,0 +1,9 @@
+// RUN: %clang_cc1  -fprofile-instrument=clang -fcoverage-mapping -emit-llvm -o - %s | FileCheck %s
+// RUN: %clang_cc1 -mllvm -profile-correlate=binary -fprofile-instrument=clang -fcoverage-mapping -emit-llvm -o - %s | FileCheck %s --check-prefix=BIN-CORRELATE
```
Specify an explicit `-target-triple` to test a COMDAT triple (e.g. Linux) and a non-COMDAT triple (e.g. macOS).

Change `// CHECK: @__llvm_profile_raw_version = {{.*}} i64 9` to test more properties including the visibility and `comdat`

MaskRay commented on llvm/docs/CommandGuide/llvm-profdata.rst:
```diff
@@ -195,8 +195,14 @@ OPTIONS
  Specify the executable or ``.dSYM`` that contains debug info for the raw profile.
- When ``-debug-info-correlate`` was used for instrumentation, use this option
- to correlate the raw profile.
+ When ``-profile-correlate=debug-info`` was used for instrumentation, use this
```
`--profile-correlate=debug-info` (double-dash)

MaskRay commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -1341,20 +1344,26 @@ void InstrProfiling::createDataVariable(InstrProfCntrInstBase *Inc,
-  if (BitmapPtr != nullptr) {
-    RelativeBitmapPtr =
-        ConstantExpr::getSub(ConstantExpr::getPtrToInt(BitmapPtr, IntPtrTy),
+  // By default counter ptr and bitmap ptr are address relative to data section.
```
This comment duplicates a bit of "// Reference the counter variable with a label difference (link-time".

We can just say: With binary profile correlation, profd is not loaded into memory. profd must reference profc with an absolute relocation.

MaskRay commented on llvm/lib/ProfileData/InstrProfCorrelator.cpp:
```diff
@@ -46,14 +73,38 @@ const char *InstrProfCorrelator::NumCountersAttributeName = "Num Counters";
+  // In COFF object file, there's a null byte at the beginning of the counter
+  // section which doesn't exist in raw profile.
+  if (Obj.getTripleObjectFormat() == Triple::COFF)
+    C->CountersSectionStart++;
```
pre-increment https://llvm.org/docs/CodingStandards.html#prefer-preincrement

MaskRay commented on llvm/test/Instrumentation/InstrProfiling/binary-correlate.ll:
```diff
@@ -0,0 +1,11 @@
+; RUN: opt < %s -passes=instrprof -profile-correlate=binary -S | FileCheck %s
```
This isn't clear how `__profd_foo` is different from `-profile-correlate=none`. This test can be added to `coverage.ll` as a new RUN line

MaskRay commented on llvm/tools/llvm-profdata/llvm-profdata.cpp:
```diff
@@ -1331,6 +1336,18 @@ static int merge_main(int argc, const char *argv[]) {
 
   cl::ParseCommandLineOptions(argc, argv, "LLVM profile data merger\n");
+  if (!DebugInfoFilename.empty() && !BinaryFilename.empty()) {
+    exitWithError("Expected only one of -debug-info, -binary-file");
```
drop braces https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

MaskRay commented on llvm/tools/llvm-profdata/llvm-profdata.cpp:
```diff
@@ -1331,6 +1336,18 @@ static int merge_main(int argc, const char *argv[]) {
+  if (!DebugInfoFilename.empty() && !BinaryFilename.empty()) {
+    exitWithError("Expected only one of -debug-info, -binary-file");
+  }
+  std::string CorrelateFilename = "";
```
remove `= ""`

MaskRay commented on compiler-rt/test/profile/instrprof-binary-correlate.c:
```diff
@@ -0,0 +1,46 @@
+// REQUIRES: linux || windows
```
Add `// REQUIRES: lld-available`

MaskRay commented on compiler-rt/test/profile/instrprof-binary-correlate.c:
```diff
@@ -0,0 +1,46 @@
+// RUN: llvm-cov show --instr-profile=%t.normal.profdata %t.normal > %t.normal.show
+
+// With -profile-correlate=binary flag
+// RUN: %clang -o %t-1.exe -fprofile-instr-generate -fcoverage-mapping -mllvm -profile-correlate=binary -fuse-ld=lld %S/Inputs/instrprof-debug-info-correlate-main.cpp %S/Inputs/instrprof-debug-info-correlate-foo.cpp
```
You can omit `.exe`. Otherwise, I suspect the file will be named `%t-1.exe.exe` on Windows.

MaskRay reviewed (commented):
The comment `The data and names sections are omitted in lightweight mode.` in compiler-rt should be updated since binary correlation is different from the lightweight mode

MaskRay commented:
The patch does not apply cleanly at HEAD. The fix-up commit `resolve conflict` contains the rebased part and a lot of unrelated changes. I think in this case, it's cleaner to squash all the commits and force push to `binary-correlate`.

(Force push is fine with me: https://discourse.llvm.org/t/force-push-and-rebase/73766/10)

ZequanWu commented:
There are some discussions about this outside the PR:
https://github.com/llvm/llvm-project/pull/70856#issuecomment-1791465183
https://discourse.llvm.org/t/rfc-add-binary-profile-correlation-to-not-load-profile-metadata-sections-into-memory-at-runtime/74565/8

Some further discussion/pre-work needed before proceeding with this PR.

ZequanWu committed:
Merge remote-tracking branch 'origin/main' into binary-correlate

ZequanWu committed:
Updated to use new section names `__llvm_covdata` and `__llvm_covnames` (or `.lprfcovdata` and `.lprfcovnames` on COFF) under binary correlation mode.

The reason is that we don't want to rely on the bit in llvm profile version variable to indicate name/data sections are empty or not: https://github.com/llvm/llvm-project/pull/71996. And simply making __llvm_prf_{data}/{names} sections not allocated in ELF does not make `&__stop___llvm_prf_data - &__start___llvm_prf_data == 0`. It still returns the size of the non-allocated sections. So, using different section names eliminate the need to perform runtime check for data/name section size.

With that, there is no need to have VARIANT_MASK_BIN_CORRELATE bit. Clang side codegen changes are not longer necessary.

ZequanWu committed:
fixup! format

ZequanWu reviewed (commented):
>  "binary" is ambiguous. I wonder whether object file correlation is better. llvm-symbolizer has an option --obj=xxx.

llvm-symbolizer's `--obj` could take an pre-linking object file. But here we need to take post-linked binary for merging.

ZequanWu committed:
Fix coff. Section names should be no greater than 8 bytes

ZequanWu commented:
> Some further discussion/pre-work needed before proceeding with this PR.

These two discussions are addressed and this is ready to be reviewed.

1. Support for merging with multiple correlation files will be done in a separate change.
2. Removed the `VARIANT_MASK_BIN_CORRELATE` and runtime check for it to determine if name/data sections should be skipped. Instead, new section names are used (`__llvm_covdata` and `__llvm_covnames`).

ZequanWu commented:
Ping.

petrhosek commented on compiler-rt/include/profile/InstrProfData.inc:
```diff
@@ -702,6 +708,8 @@ serializeValueProfDataFrom(ValueProfRecordClosure *Closure,
 #define INSTR_PROF_COVMAP_COMMON __llvm_covmap
 #define INSTR_PROF_COVFUN_COMMON __llvm_covfun
 #define INSTR_PROF_ORDERFILE_COMMON __llvm_orderfile
```
This just a nit, but could you move `orderfile` here and elsewhere to the bottom so that all `__llvm_cov*` entries are next to each other for easier readability.

ZequanWu replied:
Done.

petrhosek commented:
Can you break up all the changes to tests that replace `-debug-info-correlate` with `--profile-correlate=debug-info` into a separate PR to reduce the size of this PR?

ZequanWu committed:
re-order macros in InstrProfData.inc so __llvm_cov* are together and reverted changes in tests that use debug info correlate

ZequanWu commented:
> Can you break up all the changes to tests that replace `-debug-info-correlate` with `--profile-correlate=debug-info` into a separate PR to reduce the size of this PR?

Done.

david-xl commented:
> @david-xl , Zequan posted an [RFC](https://discourse.llvm.org/t/rfc-add-binary-profile-correlation-to-not-load-profile-metadata-sections-into-memory-at-runtime/74565/8) for this. Is there a PGO tag, or something we can use to increase visibility for PGO reviewers? I think most of the complexity for this new mode is in the PGO code, so I think the best person to review it would be someone with an interest in PGO.
> 
> I'll tag some folks, but let me know if there is a better way: @kparzysz @snehasish

I am aware of this patch and it mostly look good to me. Since there is a long list of reviewers attached to this PR, better give more time for them to chime in.  I am quite busy early part of this week, but will try to give it another round by the end of the week if no one else beats me to it.

MaskRay commented:
We have this team: @llvm/pr-subscribers-pgo

ZequanWu committed:
Merge branch 'main' into binary-correlate

david-xl reviewed (approved):
LGTM. Thanks for the new feature. (Wait a few days in case other reviewers have additional comments).

ZequanWu commented on clang/lib/CodeGen/CoverageMappingGen.cpp:
```diff
@@ -1829,6 +1833,22 @@ void CoverageMappingModuleGen::emit() {
                              llvm::GlobalValue::InternalLinkage, NamesArrVal,
                              llvm::getCoverageUnusedNamesVarName());
   }
+  const StringRef VarName(INSTR_PROF_QUOTE(INSTR_PROF_RAW_VERSION_VAR));
```
I already reverted changes in Clang because `VARIANT_MASK_BIN_CORRELATE` flag is no longer necessary: https://github.com/llvm/llvm-project/pull/69493#issuecomment-1815324995

ilovepi commented:
I think our Linux CI is seeing failures related to this patch:
https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-linux-arm64/b8761705440671462977/overview

 Profile-aarch64 :: instrprof-binary-correlate.c is failing on that bot for both x86_64 and arm64.

ilovepi commented:
Test output:
```
******************** TEST 'Profile-aarch64 :: instrprof-binary-correlate.c' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
Binary files /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal.profdata and /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.profdata differ

--
Command Output (stderr):
--
RUN: at line 3: /b/s/w/ir/x/w/llvm_build/./bin/clang    -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta   -ldl -o /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal -fprofile-instr-generate -fcoverage-mapping /b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/test/profile/Inputs/instrprof-debug-info-correlate-main.cpp /b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/test/profile/Inputs/instrprof-debug-info-correlate-foo.cpp
+ /b/s/w/ir/x/w/llvm_build/./bin/clang -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta -ldl -o /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal -fprofile-instr-generate -fcoverage-mapping /b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/test/profile/Inputs/instrprof-debug-info-correlate-main.cpp /b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/test/profile/Inputs/instrprof-debug-info-correlate-foo.cpp
RUN: at line 4: env LLVM_PROFILE_FILE=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.profraw  /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal
+ env LLVM_PROFILE_FILE=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.profraw /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal
RUN: at line 5: llvm-profdata merge -o /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal.profdata /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.profraw
+ llvm-profdata merge -o /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal.profdata /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.profraw
RUN: at line 6: llvm-cov report --instr-profile=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal.profdata /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal > /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal.report
+ llvm-cov report --instr-profile=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal.profdata /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal
RUN: at line 7: llvm-cov show --instr-profile=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal.profdata /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal > /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal.show
+ llvm-cov show --instr-profile=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal.profdata /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal
RUN: at line 10: /b/s/w/ir/x/w/llvm_build/./bin/clang    -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta   -ldl -o /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.exe -fprofile-instr-generate -fcoverage-mapping -mllvm -profile-correlate=binary /b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/test/profile/Inputs/instrprof-debug-info-correlate-main.cpp /b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/test/profile/Inputs/instrprof-debug-info-correlate-foo.cpp
+ /b/s/w/ir/x/w/llvm_build/./bin/clang -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta -ldl -o /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.exe -fprofile-instr-generate -fcoverage-mapping -mllvm -profile-correlate=binary /b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/test/profile/Inputs/instrprof-debug-info-correlate-main.cpp /b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/test/profile/Inputs/instrprof-debug-info-correlate-foo.cpp
RUN: at line 11: env LLVM_PROFILE_FILE=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.profraw  /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.exe
+ env LLVM_PROFILE_FILE=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.profraw /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.exe
RUN: at line 12: llvm-profdata merge -o /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.profdata --binary-file=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.exe /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.profraw
+ llvm-profdata merge -o /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.profdata --binary-file=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.exe /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.profraw
RUN: at line 13: llvm-cov report --instr-profile=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.profdata /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.exe > /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.report
+ llvm-cov report --instr-profile=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.profdata /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.exe
RUN: at line 14: llvm-cov show --instr-profile=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.profdata /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.exe > /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.show
+ llvm-cov show --instr-profile=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.profdata /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.exe
RUN: at line 15: diff /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal.profdata /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.profdata
+ diff /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp.normal.profdata /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-aarch64-unknown-linux-gnu-bins/compiler-rt/test/profile/Profile-aarch64/Output/instrprof-binary-correlate.c.tmp-1.profdata

--

********************
```

Log: https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8761705440671462977/+/u/clang/test/stdout

ilovepi commented:
Seems like a mismatch on the diff, so maybe the check is too stringent.

If this will take a while to fix, would you mind reverting until it can be addressed?

ZequanWu commented:
> Seems like a mismatch on the diff, so maybe the check is too stringent.
> 
> If this will take a while to fix, would you mind reverting until it can be addressed?

It passed for me locally on x64. Maybe I should use `diff <(llvm-profdata show --all-functions --counts %t.normal.profdata) <(llvm-profdata show --all-functions --counts %t.d4.profdata)` like it does on debug-info correlate test:
https://github.com/llvm/llvm-project/blob/29e043cb5c2efaad7fb203fb8240a91b77ca0c5b/compiler-rt/test/profile/Linux/instrprof-debug-info-correlate.c#L10

Sent a possible fix at: https://github.com/llvm/llvm-project/commit/f34325307eb36d6032ccea3773e3e0c1746b7f98, hope that fixes it.

ilovepi commented:
Our next cI run should finish in about 25 minutes, so we'll find out soon. https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-linux-x64/b8761696932167687409/overview

ilovepi commented:
Well, seems like someone broke ToT w/ a compiler error. I'll let you know if the forward fix fails to address the issue.

ZequanWu commented:
> Well, seems like someone broke ToT w/ a compiler error. I'll let you know if the forward fix fails to address the issue.

The latest build passed: https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-linux-x64/b8761696377585255057/overview.

ilovepi commented:
Indeed. Thank you!
